src/main/kotlin/uk/gov/communities/prsdb/webapp/clients/EpcRegisterClient.kt

```
@@ -0,0 +1,72 @@
+        return searchFor("uprn", "UPRN-$query")!!
+    }
+
+    fun getByRrn(rrn: String): String {
+        val query = URLEncoder.encode(rrn, "UTF-8")
+        return searchFor("rrn", query)!!
+    }
+
+    private fun searchFor(
+        parameterName: String,
+        parameterValue: String,
+    ): String? {
+        var errorString: String? = null
+        val bodyString =
+            client
+                .get()
+                .uri(
+                    UriComponentsBuilder
+                        .fromUriString("/api/prsdatabase/assessments/search")
+                        .queryParam(parameterName, parameterValue)
+                        .build()
+                        .toUri(),
+                ).retrieve()
+                // This ensures that we naively treat NOT_FOUND responses as valid, for searches that do not
+                // match a property/certificate. This silently ignores the error and the response body is treated
+                // like a successful response.
+                .onStatus({ it == HttpStatus.NOT_FOUND }) { request, response -> }
+                // This parses the response body and depending on the content either sets a fallback value as a side
+                // effect or throws a custom exception. We read the body stream without closing it as it will be read
+                // by the `body<String>` below.
+                .onStatus({ it == HttpStatus.BAD_REQUEST }) { request, response ->
+                    val bodyString = response.body.reader().readText()
+                    val code = getErrorCode(bodyString)
+                    if (code == "INVALID_REQUEST" || code == "BAD_REQUEST") {
+                        errorString = bodyString
+                    } else {
+                        throw PrsdbWebException("Example fallback error: $bodyString")
+                    }
+                }.body<String>()
+
+        return bodyString ?: errorString
```

Are there real cases where we'd want to return the underlying error or is this just for the purposes of the example?

---

This is fundementally just for the purposes of the example - in real life we probably wouldn't want to return the body string either. This is basically just showing how to extract info out of the `onStatus` without throwing an exception.